File: v1/examples/python/06_lpr_search.py
"""
06_lpr_search.py - TS-API v1 LPR (License Plate Recognition)

Endpoints:
  GET /api/v1/lpr/source
      - List LPR-enabled sources (cameras)

  GET /api/v1/lpr/log?timeBegin=...&timeEnd=...
      - Search LPR recognition log (timeBegin and timeEnd are required)

  GET /api/v1/lpr/similar?keyword=...&timeBegin=...&timeEnd=...
      - Search for similar plate numbers (fuzzy match)

  GET /api/v1/lpr/log?timeBegin=...&timeEnd=...&export=true
      - Export LPR log as downloadable file (CSV/Excel)

WARNING: Exporting large datasets (10,000+ records) may cause HTTP timeout
  errors. For bulk exports, narrow the time range or use pagination
  (at/maxCount) to keep each request under a manageable size.
"""

import os
import sys
from datetime import datetime, timedelta

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import NVR_API_KEY
from http_client import NvrClient

if not NVR_API_KEY:
    print('NVR_API_KEY environment variable is required')
    sys.exit(1)


def main():
    client = NvrClient()
    client.set_api_key(NVR_API_KEY)

    # --- LPR Sources ---
    print("=== LPR Sources ===")
    r = client.get('/api/v1/lpr/source')
    if r.status_code == 200:
        sources = r.json()
        for s in sources:
            print(f"  {s}")
    else:
        print(f"  Status: {r.status_code}")

    # --- LPR Log (last 24 hours) ---
    # timeBegin and timeEnd are REQUIRED parameters
    now = datetime.now()
    time_end = now.strftime('%Y-%m-%d %H:%M:%S')
    time_begin = (now - timedelta(hours=24)).strftime('%Y-%m-%d %H:%M:%S')

    print(f"\n=== LPR Log ({time_begin} ~ {time_end}) ===")
    r = client.get('/api/v1/lpr/log', params={
        'timeBegin': time_begin,
        'timeEnd': time_end,
    })
    if r.status_code == 200:
        records = r.json().get('data', [])
        print(f"  {len(records)} records found")
        for rec in records[:5]:  # Show first 5
            print(f"  [{rec.get('timeRange', '')}] plateNo={rec.get('plateNo', '')} "
                  f"ch={rec.get('chid', '')}")
    else:
        print(f"  Status: {r.status_code}")

    # --- Similar Plate Search ---
    plate_query = '1234'  # Partial or similar plate number
    print(f"\n=== Similar Plate Search (keyword={plate_query}) ===")
    r = client.get('/api/v1/lpr/similar', params={
        'keyword': plate_query,
        'timeBegin': time_begin,
        'timeEnd': time_end,
    })
    if r.status_code == 200:
        body = r.json()
        # similar returns flat array of plate strings, or {data:[...]} wrapper
        results = body.get('data', body) if isinstance(body, dict) else body
        print(f"  {len(results)} similar plates found")
        for rec in results[:10]:
            if isinstance(rec, str):
                print(f"    {rec}")
            else:
                print(f"  [{rec.get('timeRange', '')}] plateNo={rec.get('plateNo', '')}")
    else:
        print(f"  Status: {r.status_code}")

    # --- Export LPR Log ---
    print(f"\n=== Export LPR Log ===")
    r = client.get('/api/v1/lpr/log', params={
        'timeBegin': time_begin,
        'timeEnd': time_end,
        'export': 'true',
    })
    if r.status_code == 200:
        content_type = r.headers.get('Content-Type', '')
        print(f"  Content-Type: {content_type}")
        print(f"  Size: {len(r.content)} bytes")
        # Save to file:
        # with open('lpr_export.csv', 'wb') as f:
        #     f.write(r.content)
    else:
        print(f"  Status: {r.status_code}")


if __name__ == '__main__':
    main()

File: v1/examples/csharp/06_LprSearch.cs
// 06_LprSearch.cs - v1 API License Plate Recognition Search Example
//
// Endpoints:
//   GET /api/v1/lpr/source                                            - LPR source list
//   GET /api/v1/lpr/log?timeBegin=...&timeEnd=...                     - Plate number search
//   GET /api/v1/lpr/similar?keyword=...&timeBegin=...&timeEnd=...     - Similar plate search
//
// Note: timeBegin and timeEnd are REQUIRED parameters.
//
// WARNING: Exporting large datasets (10,000+ records) may cause HTTP timeout
//   errors. For bulk exports, narrow the time range or use pagination
//   (at/maxCount) to keep each request under a manageable size.

using System;
using System.Text.Json;
using System.Threading.Tasks;

namespace TsApiExamples.V1
{
    class LprSearch
    {
        static async Task Main(string[] args)
        {
            using var client = new NvrClient();
            if (string.IsNullOrEmpty(client.ApiKey))
            {
                Console.WriteLine("NVR_API_KEY environment variable is required");
                return;
            }
            client.SetApiKey(client.ApiKey);

            // ── Get LPR sources ──
            Console.WriteLine("=== LPR Sources ===");
            var sr = await client.GetAsync("/api/v1/lpr/source");
            if (!sr.IsSuccessStatusCode)
            {
                Console.WriteLine($"  Failed: {(int)sr.StatusCode}");
                return;
            }

            using var sdoc = await NvrClient.ReadJsonAsync(sr);
            foreach (var src in sdoc.RootElement.EnumerateArray())
            {
                Console.WriteLine($"  {src}");
            }

            // ── Search plate numbers ──
            var now = DateTime.Now;
            var weekAgo = now.AddDays(-7);
            var timeBegin = weekAgo.ToString("yyyyMMddHHmmss");
            var timeEnd = now.ToString("yyyyMMddHHmmss");

            Console.WriteLine($"\n=== Plate Number Search ===");
            Console.WriteLine($"  Period: {weekAgo:yyyy-MM-dd} ~ {now:yyyy-MM-dd}");

            var lr = await client.GetAsync(
                $"/api/v1/lpr/log?timeBegin={timeBegin}&timeEnd={timeEnd}");
            if (!lr.IsSuccessStatusCode)
            {
                Console.WriteLine($"  Failed: {(int)lr.StatusCode}");
                return;
            }

            // Response: {totalCount: N, at: N, data: [...]}
            using var ldoc = await NvrClient.ReadJsonAsync(lr);
            var lroot = ldoc.RootElement;
            var results = lroot.TryGetProperty("data", out var dataProp)
                ? dataProp : lroot;
            int totalCount = lroot.TryGetProperty("totalCount", out var tcProp)
                ? tcProp.GetInt32() : results.GetArrayLength();
            Console.WriteLine($"  Results: {totalCount}");

            int shown = 0;
            foreach (var item in results.EnumerateArray())
            {
                if (++shown > 5) break;
                Console.WriteLine($"  [{shown}] {item}");
            }
            if (totalCount > 5)
                Console.WriteLine($"  ... {totalCount - 5} more");

            // ── Similar plate search ──
            string keyword = args.Length > 0 ? args[0] : "1234";
            Console.WriteLine($"\n=== Similar Plate Search (keyword: {keyword}) ===");

            var smr = await client.GetAsync(
                $"/api/v1/lpr/similar?keyword={keyword}&timeBegin={timeBegin}&timeEnd={timeEnd}");
            if (!smr.IsSuccessStatusCode)
            {
                Console.WriteLine($"  Failed: {(int)smr.StatusCode}");
                return;
            }

            // Response: flat array of plate strings (e.g. ["12A3456", "78B9012"])
            using var smdoc = await NvrClient.ReadJsonAsync(smr);
            var simRoot = smdoc.RootElement;
            var simResults = simRoot.ValueKind == JsonValueKind.Array
                ? simRoot
                : simRoot.TryGetProperty("data", out var simData) ? simData : simRoot;
            Console.WriteLine($"  Results: {simResults.GetArrayLength()}");

            shown = 0;
            foreach (var item in simResults.EnumerateArray())
            {
                if (++shown > 5) break;
                Console.WriteLine($"  [{shown}] {item}");
            }
        }
    }
}
